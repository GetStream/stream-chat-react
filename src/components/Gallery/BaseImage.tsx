import React, { forwardRef, useEffect, useState } from 'react';
import clsx from 'clsx';
import { DownloadButton } from '../Attachment';

export type BaseImageProps = React.ComponentPropsWithRef<'img'>;

export const BaseImage = forwardRef<HTMLImageElement, BaseImageProps>(function BaseImage(
  { ...props },
  ref,
) {
  const { className: propsClassName, onError: propsOnError } = props;
  const [error, setError] = useState(false);

  useEffect(
    () => () => {
      setError(false);
    },
    [props.src],
  );

  return (
    <>
      <img
        data-testid='str-chat__base-image'
        {...props}
        className={clsx(propsClassName, 'str-chat__base-image', {
          'str-chat__base-image--load-failed': error,
        })}
        onError={(e) => {
          setError(true);
          propsOnError?.(e);
        }}
        ref={ref}
      />
      {error && <DownloadButton assetUrl={props.src} />}
    </>
  );
});
