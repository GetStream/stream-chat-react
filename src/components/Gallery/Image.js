/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';

import ModalWrapper from './ModalWrapper';
import { sanitizeUrl } from '@braintree/sanitize-url';

/**
 * Image - Small wrapper around an image tag, supports thumbnails
 *
 * @example ../../docs/Image.md
 * @extends {React.PureComponent<import('type').ImageProps>}
 */
class Image extends React.PureComponent {
  static propTypes = {
    /** The full size image url */
    image_url: PropTypes.string,
    /** The thumb url */
    thumb_url: PropTypes.string,
    /** The text fallback for the image */
    fallback: PropTypes.string,
  };
  state = {
    modalIsOpen: false,
    currentIndex: 0,
  };

  toggleModal = () => {
    this.setState((state) => ({
      modalIsOpen: !state.modalIsOpen,
    }));
  };

  render() {
    const { image_url, thumb_url, fallback } = this.props;
    const imageSrc = sanitizeUrl(image_url || thumb_url);
    const formattedArray = [{ src: imageSrc }];
    return (
      <React.Fragment>
        <img
          className="str-chat__message-attachment--img"
          onClick={this.toggleModal}
          src={imageSrc}
          alt={fallback}
          data-testid="image-test"
        />

        <ModalWrapper
          images={formattedArray}
          toggleModal={this.toggleModal}
          index={this.state.currentIndex}
          modalIsOpen={this.state.modalIsOpen}
        />
      </React.Fragment>
    );
  }
}

export default Image;
