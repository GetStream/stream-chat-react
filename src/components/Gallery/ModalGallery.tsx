import React, { useMemo } from 'react';
import ImageGallery from 'react-image-gallery';

import type { Attachment } from 'stream-chat';
import type { DefaultStreamChatGenerics } from '../../types/types';

export type ModalGalleryProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  /** The images for the Carousel component */
  images: Attachment<StreamChatGenerics>[];
  /** The index for the component */
  index?: number;
};

export const ModalGallery = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: ModalGalleryProps<StreamChatGenerics>,
) => {
  const { images, index } = props;

  const formattedArray = useMemo(
    () =>
      images.map((image) => {
        const imageSrc = image.image_url || image.thumb_url || '';
        return {
          original: imageSrc,
          originalAlt: 'User uploaded content',
          source: imageSrc,
        };
      }),
    [images],
  );

  return (
    <ImageGallery
      items={formattedArray}
      showIndex={true}
      showPlayButton={false}
      showThumbnails={false}
      startIndex={index}
    />
  );
};
