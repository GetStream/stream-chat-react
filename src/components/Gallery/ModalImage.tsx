import React from 'react';

export type ModalImageProps = {
  /** The src attribute for the image element */
  data: { source: string };
};

export const ModalImage: React.FC<ModalImageProps> = (props) => {
  const { data } = props;
  return (
    <div className='str-chat__modal-image__wrapper' data-testid='modal-image'>
      <img alt='User uploaded content' className='str-chat__modal-image__image' src={data.source} />
    </div>
  );
};
