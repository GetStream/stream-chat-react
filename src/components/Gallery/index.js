export { default as Gallery } from './Gallery';
export { default as ImageComponent } from './Image';
export { default as ModalImage } from './ModalImage';
export { default as ModalWrapper } from './ModalWrapper';
