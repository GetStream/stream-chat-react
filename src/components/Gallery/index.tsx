export * from './Gallery';
export * from './Image';
export * from './ModalWrapper';
