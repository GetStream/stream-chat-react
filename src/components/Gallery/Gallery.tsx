import React, { CSSProperties, MutableRefObject, useState } from 'react';
import { sanitizeUrl } from '@braintree/sanitize-url';
import clsx from 'clsx';

import { Modal } from '../Modal';
import { ModalGallery as DefaultModalGallery } from './ModalGallery';

import { useComponentContext } from '../../context/ComponentContext';
import { useTranslationContext } from '../../context/TranslationContext';

import type { Attachment } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../types/types';

export type GalleryProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  images: ((
    | {
        image_url?: string | undefined;
        thumb_url?: string | undefined;
      }
    | Attachment<StreamChatGenerics>
  ) & { previewUrl?: string; style?: CSSProperties })[];
  innerRefs?: MutableRefObject<(HTMLElement | null)[]>;
};

const UnMemoizedGallery = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: GalleryProps<StreamChatGenerics>,
) => {
  const { images, innerRefs } = props;

  const [index, setIndex] = useState(0);
  const [modalOpen, setModalOpen] = useState(false);

  const { ModalGallery = DefaultModalGallery } = useComponentContext('Gallery');
  const { t } = useTranslationContext('Gallery');

  const countImagesDisplayedInPreview = 4;
  const lastImageIndexInPreview = countImagesDisplayedInPreview - 1;

  const toggleModal = (selectedIndex: number) => {
    if (modalOpen) {
      setModalOpen(false);
    } else {
      setIndex(selectedIndex);
      setModalOpen(true);
    }
  };

  const renderImages = images.slice(0, countImagesDisplayedInPreview).map((image, i) =>
    i === lastImageIndexInPreview && images.length > countImagesDisplayedInPreview ? (
      <button
        className='str-chat__gallery-placeholder'
        data-testid='gallery-image-last'
        key={`gallery-image-${i}`}
        onClick={() => toggleModal(i)}
        style={{
          backgroundImage: `url(${
            images[lastImageIndexInPreview].previewUrl ||
            images[lastImageIndexInPreview].image_url ||
            images[lastImageIndexInPreview].thumb_url
          })`,
          ...image.style,
        }}
        {...(innerRefs?.current && { ref: (r) => (innerRefs.current[i] = r) })}
      >
        <p>
          {t<string>('{{ imageCount }} more', {
            imageCount: images.length - countImagesDisplayedInPreview,
          })}
        </p>
      </button>
    ) : (
      <button
        className='str-chat__gallery-image'
        data-testid='gallery-image'
        key={`gallery-image-${i}`}
        onClick={() => toggleModal(i)}
      >
        <img
          alt='User uploaded content'
          src={sanitizeUrl(image.previewUrl || image.image_url || image.thumb_url)}
          style={image.style}
          {...(innerRefs?.current && { ref: (r) => (innerRefs.current[i] = r) })}
        />
      </button>
    ),
  );

  const className = clsx('str-chat__gallery', {
    'str-chat__gallery--square': images.length > lastImageIndexInPreview,
    'str-chat__gallery-two-rows': images.length > 2,
  });

  return (
    <div className={className}>
      {renderImages}
      <Modal onClose={() => setModalOpen((modalOpen) => !modalOpen)} open={modalOpen}>
        <ModalGallery images={images} index={index} />
      </Modal>
    </div>
  );
};

/**
 * Displays images in a simple responsive grid with a light box to view the images.
 */
export const Gallery = React.memo(UnMemoizedGallery) as typeof UnMemoizedGallery;
