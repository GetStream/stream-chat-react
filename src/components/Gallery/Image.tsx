import React, { CSSProperties, MutableRefObject, useState } from 'react';
import { sanitizeUrl } from '@braintree/sanitize-url';

import { Modal } from '../Modal';
import { ModalGallery as DefaultModalGallery } from './ModalGallery';
import { useComponentContext } from '../../context';

import type { Attachment } from 'stream-chat';
import type { DefaultStreamChatGenerics, Dimensions } from '../../types/types';

export type ImageProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  dimensions?: Dimensions;
  innerRef?: MutableRefObject<HTMLImageElement | null>;
  previewUrl?: string;
  style?: CSSProperties;
} & (
  | {
      /** The text fallback for the image */
      fallback?: string;
      /** The full size image url */
      image_url?: string;
      /** The thumb url */
      thumb_url?: string;
    }
  | Attachment<StreamChatGenerics>
);

/**
 * A simple component that displays an image.
 */
export const ImageComponent = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: ImageProps<StreamChatGenerics>,
) => {
  const { dimensions = {}, fallback, image_url, thumb_url, innerRef, previewUrl, style } = props;

  const [modalIsOpen, setModalIsOpen] = useState(false);
  const { ModalGallery = DefaultModalGallery } = useComponentContext('ImageComponent');

  const imageSrc = sanitizeUrl(previewUrl || image_url || thumb_url);

  const toggleModal = () => setModalIsOpen((modalIsOpen) => !modalIsOpen);

  return (
    <>
      <img
        alt={fallback}
        className='str-chat__message-attachment--img'
        data-testid='image-test'
        onClick={toggleModal}
        src={imageSrc}
        style={style}
        tabIndex={0}
        {...dimensions}
        {...(innerRef && { ref: innerRef })}
      />
      <Modal onClose={toggleModal} open={modalIsOpen}>
        <ModalGallery images={[props]} index={0} />
      </Modal>
    </>
  );
};
