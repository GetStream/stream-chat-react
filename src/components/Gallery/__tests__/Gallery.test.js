import React from 'react';
import { nanoid } from 'nanoid';
import renderer from 'react-test-renderer';
import { act, cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { getTestClientWithUser } from '../../../mock-builders';

import { Chat } from '../../Chat';
import { Gallery } from '../Gallery';

import { ComponentProvider } from '../../../context/ComponentContext';

let chatClient;

const mockGalleryAssets = [
  {
    original: 'https://placeimg.com/640/480/any',
    originalAlt: 'User uploaded content',
    src: 'https://placeimg.com/640/480/any',
  },
  {
    original: 'https://placeimg.com/640/480/any',
    originalAlt: 'User uploaded content',
    src: 'https://placeimg.com/640/480/any',
  },
  {
    original: 'https://placeimg.com/640/480/any',
    originalAlt: 'User uploaded content',
    src: 'https://placeimg.com/640/480/any',
  },
  {
    original: 'https://placeimg.com/640/480/any',
    originalAlt: 'User uploaded content',
    src: 'https://placeimg.com/640/480/any',
  },
  {
    original: 'https://placeimg.com/640/480/any',
    originalAlt: 'User uploaded content',
    src: 'https://placeimg.com/640/480/any',
  },
];

describe('Gallery', () => {
  afterEach(cleanup);

  it('should render component with default props', () => {
    const tree = renderer
      .create(
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets.slice(0, 2)} />
        </ComponentProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render component with 3 images', () => {
    const tree = renderer
      .create(
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets.slice(0, 3)} />
        </ComponentProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render component with 4 images', () => {
    const tree = renderer
      .create(
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets.slice(0, 4)} />
        </ComponentProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render component with 5 images', () => {
    const tree = renderer
      .create(
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets} />
        </ComponentProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should open modal on image click', async () => {
    jest.spyOn(console, 'warn').mockImplementation(() => null);

    const { getByTestId, getByTitle } = render(
      <ComponentProvider value={{}}>
        <Gallery images={mockGalleryAssets.slice(0, 1)} />
      </ComponentProvider>,
    );

    fireEvent.click(getByTestId('gallery-image'));

    await waitFor(() => {
      expect(getByTitle('Close')).toBeInTheDocument();
    });
  });

  it('should display correct image count', async () => {
    chatClient = await getTestClientWithUser({ id: 'test' });
    const { getByText } = await render(
      <Chat client={chatClient}>
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets} />
        </ComponentProvider>
      </Chat>,
    );
    await waitFor(() => {
      expect(getByText('1 more')).toBeInTheDocument();
    });
  });

  it('should open the modal with image displayed under the "X more" overlay if clicked on the overlay', async () => {
    chatClient = await getTestClientWithUser({ id: 'test' });
    const { container, getByText } = render(
      <Chat client={chatClient}>
        <ComponentProvider value={{}}>
          <Gallery images={mockGalleryAssets} />
        </ComponentProvider>
      </Chat>,
    );

    const overlay = await waitFor(() => getByText('1 more'));
    act(() => {
      fireEvent.click(overlay);
    });

    await waitFor(() => {
      expect(container.querySelector('.image-gallery-index')).toHaveTextContent('4 / 5');
    });
  });

  it('should render custom ModalGallery component from context', async () => {
    const galleryContent = nanoid();
    const CustomGallery = () => <div>{galleryContent}</div>;
    const { getAllByTestId, getByText } = render(
      <ComponentProvider value={{ ModalGallery: CustomGallery }}>
        <Gallery images={mockGalleryAssets} />
      </ComponentProvider>,
    );

    fireEvent.click(getAllByTestId('gallery-image')[0]);

    await waitFor(() => {
      expect(getByText(galleryContent)).toBeInTheDocument();
    });
  });
});
