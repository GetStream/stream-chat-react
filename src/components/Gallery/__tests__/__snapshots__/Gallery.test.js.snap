// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery should render component with 3 images 1`] = `
<div
  className="str-chat__gallery str-chat__gallery-two-rows"
>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
</div>
`;

exports[`Gallery should render component with 4 images 1`] = `
<div
  className="str-chat__gallery str-chat__gallery--square str-chat__gallery-two-rows"
>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
</div>
`;

exports[`Gallery should render component with 5 images 1`] = `
<div
  className="str-chat__gallery str-chat__gallery--square str-chat__gallery-two-rows"
>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-placeholder"
    data-testid="gallery-image-last"
    onClick={[Function]}
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  >
    <p>
      {{ imageCount }} more
    </p>
  </button>
</div>
`;

exports[`Gallery should render component with default props 1`] = `
<div
  className="str-chat__gallery"
>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
  <button
    className="str-chat__gallery-image"
    data-testid="gallery-image"
    onClick={[Function]}
  >
    <img
      alt="User uploaded content"
      src="about:blank"
    />
  </button>
</div>
`;
